lib/containers/accumulators/check-suites-accumulator.js

```diff
@@ -0,0 +1,156 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import {graphql, createPaginationContainer} from 'react-relay';
+
+import {PAGE_SIZE, PAGINATION_WAIT_TIME_MS} from '../../helpers';
+import {RelayConnectionPropType} from '../../prop-types';
+import CheckRunsAccumulator from './check-runs-accumulator';
+import Accumulator from './accumulator';
+
+export class BareCheckSuitesAccumulator extends React.Component {
+  static propTypes = {
+    // Relay
+    relay: PropTypes.shape({
+      hasMore: PropTypes.func.isRequired,
+      loadMore: PropTypes.func.isRequired,
+      isLoading: PropTypes.func.isRequired,
+    }).isRequired,
+    commit: PropTypes.shape({
+      checkSuites: RelayConnectionPropType(
+        PropTypes.object,
+      ),
+    }).isRequired,
+
+    // Render prop. Called with (array of errors, array of check suites, map of runs per suite, loading)
+    children: PropTypes.func.isRequired,
+
+    // Called right after refetch happens
+    onDidRefetch: PropTypes.func.isRequired,
+  }
+
+  render() {
+    const resultBatch = this.props.commit.checkSuites.edges.map(edge => edge.node);
+
+    return (
+      <Accumulator
+        relay={this.props.relay}
+        resultBatch={resultBatch}
+        onDidRefetch={this.props.onDidRefetch}
+        pageSize={PAGE_SIZE}
+        waitTimeMs={PAGINATION_WAIT_TIME_MS}>
+        {this.renderCheckSuites}
+      </Accumulator>
+    );
+  }
+
+  renderCheckSuites = (err, suites, loading) => {
+    if (err) {
+      return this.props.children({
+        errors: [err],
+        suites,
+        runsBySuite: new Map(),
+        loading,
+      });
+    }
+
+    return this.renderCheckSuite({errors: [], suites, runsBySuite: new Map(), loading}, suites);
+  }
+
+  renderCheckSuite(payload, suites) {
+    if (suites.length === 0) {
+      return this.props.children(payload);
+    }
+
+    const [suite] = suites;
+    return (
+      <CheckRunsAccumulator
+        onDidRefetch={this.props.onDidRefetch}
+        checkSuite={suite}>
+        {({error, checkRuns, loading: runsLoading}) => {
+          if (error) {
+            payload.errors.push(error);
+          }
+
+          payload.runsBySuite.set(suite, checkRuns);
+          payload.loading = payload.loading || runsLoading;
+          return this.renderCheckSuite(payload, suites.slice(1));
```

Comment:
They actually _do_ run in parallel! `<ChecksRunAccumulator>` doesn't wait for its results to complete before rendering its children prop. It's kind of hard to reason about, but it goes something like this:

1. GraphQL query 0 is performed by the enclosing `<QueryRenderer>`. The `checkSuitesAssumulator_commit` fragment here is included with first = 10 and cursor = null to fetch the first page of check suites, and the `checkRunsAccumulator_checkSuite` fragment is included with first = 20 and cursor = null to fetch the first page of check runs for each. A `<LoadingView>` is likely rendered before the result arrives.
2. The response from GraphQL query 0 arrives. It contains the first 10 CheckSuites and the first 20 CheckRuns for each. Now `<CheckSuiteAccumulator>` is rendered, which recursively renders 10 nested `<CheckRunAccumulator>` instaces, with those first pages of results in their respective props.
3. Suppose there's another page of 10 CheckSuites, and each CheckSuite has a second page of 20 runs within it. After a delay (100ms), GraphQL query 1 is fired from the CheckSuiteAccumulator's pagination container to fetch the second page of suites. Simultaneously, GraphQL queries 2 through 12 are fired by each CheckRunAccumulator to fetch the second page of runs for each known suite.
4. When GraphQL query 1's response arrives, the `<CheckSuiteAccumulator>` component is re-rendered with the new results, which in turn re-renders the existing nested `<CheckRunAccumulator>` children, then adds another 10 `<CheckRunAccumulator>` components for the newly arrived second page of suite results. Similarly, the responses for GraphQL queries 2 through 12 trigger re-renders of each individual `<CheckRunAccumulator>`.

This is the same strategy we ended up using for accumulating review comments within review threads (by which I mean I mostly copied the logic directly :eyes:).

What I _wish_ we could do is consolidate queries 2-12 into a single round-trip to avoid the "N+1" problem it's causing. Hopefully, the numbers of check suites and runs on real repositories are small enough that it's not an issue in practice?